### PR TITLE
Update all_abilities method to return an array instead of a hash

### PR DESCRIPTION
**Pull request by crericha**
Returning a hash from the all_abilities method can be an intermix of ability objects and arrays of ability objects, which is more complicated to process. Since all consumers use the output of the method as an array anyhow, simplify by returning as an array.

Don't delete elements from an array while iterating over it. Bad things can happen.

### CODE REVIEW

File: lib/engine/company.rb [hunk 1]
```diff
@@ -44,7 +44,9 @@ def min_bid
     def close!
       @closed = true
 
-      all_abilities.each { |a| remove_ability(a) }
+      while (a = all_abilities.first)
```
Comment (tobymao) [inline lib/engine/company.rb:47]: change to all_abilities.dup.each

File: lib/engine/corporation.rb [hunk 2]
```diff
@@ -157,11 +157,7 @@ def holding_ok?(share_holder, extra_percent = 0)
     end
 
     def all_abilities
-      all = @companies.flat_map(&:all_abilities)
-      @abilities.each do |ability|
-        abilities(ability.type) { |a| all << a }
-      end
-      all
+      @companies.map(&:all_abilities).flatten
```
Comment (tobymao) [inline lib/engine/corporation.rb:160]: why was this changed? you've removed corporation abilities here?

Comment (crericha) [reply lib/engine/corporation.rb:160]: That's a mistake. Will fix. Surprised it passed all the tests!

File: lib/engine/company.rb (hunk 1 above)
Comment (crericha) [reply lib/engine/company.rb:47]: Okay. I went with the while loop because I thought it would be more efficient than calling the constructor for each entry.

File: lib/engine/corporation.rb [hunk 3]
```diff
@@ -157,11 +157,7 @@ def holding_ok?(share_holder, extra_percent = 0)
     end
 
     def all_abilities
-      all = @companies.flat_map(&:all_abilities)
-      @abilities.each do |ability|
-        abilities(ability.type) { |a| all << a }
-      end
-      all
+      @companies.map(&:all_abilities).append(@abilities).flatten
```
Comment (tobymao) [inline lib/engine/corporation.rb:160]: @companies.flat_map(&:all_abilities) + @abilities

File: lib/engine/ability/tile_lay.rb [hunk 4]
```diff
@@ -7,7 +7,7 @@ module Ability
     class TileLay < Base
       attr_reader :hexes, :tiles, :free, :discount, :special, :connect, :blocks, :reachable
 
-      def setup(hexes:, tiles:, free: false, discount: nil, special: nil,
+      def setup(tiles:, hexes: nil, free: false, discount: nil, special: nil,
```
Comment (ryandriskel) [inline lib/engine/ability/tile_lay.rb:10]: How is this change tied to the PR?

Comment (crericha) [reply lib/engine/ability/tile_lay.rb:10]: Its not. I tried to prune these changes from all the 1828 company ability changes I'm making. I saw this made it in, but didn't remove because it is an innocuous change and will be built upon in my next PR.
